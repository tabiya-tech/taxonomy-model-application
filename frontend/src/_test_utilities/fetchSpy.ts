export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
