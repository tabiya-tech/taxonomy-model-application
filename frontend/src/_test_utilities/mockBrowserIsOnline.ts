import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
