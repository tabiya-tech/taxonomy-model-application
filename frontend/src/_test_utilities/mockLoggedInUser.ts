import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
