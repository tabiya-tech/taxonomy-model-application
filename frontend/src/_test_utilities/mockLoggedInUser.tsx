import {
  AuthContextValue,
  TabiyaUser,
  authContextDefaultValue,
  AuthContext,
  AuthProviderProps,
} from "src/auth/AuthProvider";
import AuthAPISpecs from "api-specifications/auth";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUsers.Anonymous;

  const value = {
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: AuthAPISpecs.Enums.TabiyaRoles) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues,
  };

  // @ts-ignore
  jest.spyOn(require("src/auth/AuthProvider"), "AuthProvider").mockImplementation((props: AuthProviderProps) => {
    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>;
  });
}

export const TestUsers = {
  ModelManager: {
    username: "ModelManagerUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER],
  },
  RegisteredUser: {
    username: "RegisteredUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER],
  },
  Anonymous: {
    username: "AnonymousUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS],
  },
};
