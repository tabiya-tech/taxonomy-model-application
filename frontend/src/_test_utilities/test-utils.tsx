// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
