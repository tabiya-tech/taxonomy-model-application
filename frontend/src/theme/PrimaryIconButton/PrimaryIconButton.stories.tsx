import PrimaryIconButton from "./PrimaryIconButton";
import { Meta, StoryObj } from "@storybook/react";
import { ContentCopy } from "@mui/icons-material";
import React from "react";
import Box from "@mui/material/Box";

const meta: Meta<typeof PrimaryIconButton> = {
  title: "Components/PrimaryIconButton",
  component: PrimaryIconButton,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof PrimaryIconButton>;

export const Shown: Story = {
  args: {
    children: (
      <Box sx={{ padding: "5px" }}>
        <ContentCopy />
      </Box>
    ),
    title: "Copy to clipboard",
  },
};
