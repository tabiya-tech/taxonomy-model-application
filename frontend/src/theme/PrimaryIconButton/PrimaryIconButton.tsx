import React, { useContext } from "react";
import { IconButton, IconButtonProps, styled } from "@mui/material";
import { IsOnlineContext } from "src/app/providers";
import CircleOutlinedIcon from "@mui/icons-material/CircleOutlined";

interface PrimaryIconButtonProps extends IconButtonProps {
  disableWhenOffline?: boolean;
}

const StyledIconButton = styled(IconButton)(({ theme }) => ({
  borderRadius: theme.spacing(theme.tabiyaSpacing.sm),
  lineHeight: "0",
  padding: 0,
  color: theme.palette.primary.main,
  ":hover": {
    backgroundColor: theme.palette.primary.dark,
    color: theme.palette.primary.contrastText,
  },
  ":active": {
    backgroundColor: theme.palette.primary.main,
    color: theme.palette.primary.contrastText,
  },
}));

const PrimaryIconButton: React.FC<PrimaryIconButtonProps> = ({
  children,
  disabled,
  disableWhenOffline,
  ...props
}: Readonly<PrimaryIconButtonProps>) => {
  const isOnline = useContext(IsOnlineContext);

  return (
    <StyledIconButton color={"primary"} disabled={Boolean(disabled || (disableWhenOffline && !isOnline))} {...props}>
      {children ?? <CircleOutlinedIcon sx={{ padding: 0, margin: 0 }} />}
    </StyledIconButton>
  );
};

export default PrimaryIconButton;
