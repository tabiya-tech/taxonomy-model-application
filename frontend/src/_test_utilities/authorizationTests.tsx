import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
