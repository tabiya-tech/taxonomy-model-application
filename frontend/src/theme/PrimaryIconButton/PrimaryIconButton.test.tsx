// mute the console
import "src/_test_utilities/consoleMock";

import React from "react";
import { render, screen } from "src/_test_utilities/test-utils";
import PrimaryIconButton from "./PrimaryIconButton";
import { mockBrowserIsOnLine, unmockBrowserIsOnLine } from "src/_test_utilities/mockBrowserIsOnline";

describe("PrimaryIconButton tests", () => {
  beforeEach(() => {
    unmockBrowserIsOnLine();
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  test("should render the icon button with default props", () => {
    // GIVEN a PrimaryIconButton component
    // WHEN the component is rendered
    const givenTestID = "test-icon";
    render(<PrimaryIconButton data-testid={givenTestID} />);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND expect the icon to be in the document
    const primaryIconButton = screen.getByTestId(givenTestID);
    expect(primaryIconButton).toBeInTheDocument();
    // AND to match the snapshot
    expect(primaryIconButton).toMatchSnapshot();
  });

  test("should render the icon button with provided icon", () => {
    // GIVEN a PrimaryIconButton component with a custom icon
    const customIcon = <svg data-testid={"test-icon"} />;

    // WHEN the component is rendered
    render(<PrimaryIconButton>{customIcon}</PrimaryIconButton>);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND expect the icon to be in the document
    const actualElement = screen.getByTestId("test-icon");
    expect(actualElement).toBeInTheDocument();
    // AND to match the snapshot
    expect(actualElement).toMatchSnapshot();
  });

  describe.each([
    [true, { disable: true, disableWhenOffline: true, isOnline: true }],
    [true, { disable: true, disableWhenOffline: true, isOnline: false }],
    [true, { disable: true, disableWhenOffline: false, isOnline: true }],
    [true, { disable: true, disableWhenOffline: false, isOnline: false }],
    [true, { disable: true, disableWhenOffline: undefined, isOnline: false }],
    [false, { disable: false, disableWhenOffline: true, isOnline: true }],
    [true, { disable: false, disableWhenOffline: true, isOnline: false }],
    [false, { disable: false, disableWhenOffline: false, isOnline: true }],
    [false, { disable: false, disableWhenOffline: false, isOnline: false }],
    [false, { disable: false, disableWhenOffline: undefined, isOnline: false }],
    [false, { disable: undefined, disableWhenOffline: true, isOnline: true }],
    [true, { disable: undefined, disableWhenOffline: true, isOnline: false }],
    [false, { disable: undefined, disableWhenOffline: undefined, isOnline: true }],
    [false, { disable: undefined, disableWhenOffline: undefined, isOnline: false }],
  ])("Disabled/enabled state", (expectedState, testCase) => {
    test(`should render the icon button disabled = ${expectedState} when ${JSON.stringify(testCase)}`, () => {
      mockBrowserIsOnLine(testCase.isOnline);

      // WHEN the component is rendered
      render(
        <PrimaryIconButton
          disabled={testCase.disable}
          disableWhenOffline={testCase.disableWhenOffline}
          data-testid={"test-icon"}
        />
      );

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
      // AND the icon button to be enabled
      const primaryIconButton = screen.getByTestId("test-icon");
      expect(primaryIconButton).toHaveProperty("disabled", expectedState);
    });
  });

  test("should render enable->disabled->enabled when online status changes", async () => {
    // GIVEN that the internet status is online
    mockBrowserIsOnLine(true);

    // GIVEN the component is rendered
    render(<PrimaryIconButton disableWhenOffline={true} data-testid={"test-icon"} />);

    // THEN expect the icon button to be enabled
    const primaryIconButton = screen.getByTestId("test-icon");
    expect(primaryIconButton).toBeEnabled();

    // WHEN the internet status changes to offline
    mockBrowserIsOnLine(false);

    // THEN expect the icon button to be disabled
    expect(primaryIconButton).toBeDisabled();

    // WHEN the internet status changes back to online
    mockBrowserIsOnLine(true);

    // THEN expect the icon button to be enabled
    expect(primaryIconButton).toBeEnabled();
  });
});
