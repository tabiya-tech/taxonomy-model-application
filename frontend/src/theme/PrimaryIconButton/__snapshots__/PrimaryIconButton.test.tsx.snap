// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrimaryIconButton tests should render the icon button with default props 1`] = `
<button
  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1e844z7-MuiButtonBase-root-MuiIconButton-root"
  data-testid="test-icon"
  tabindex="0"
  type="button"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1x821ac-MuiSvgIcon-root"
    data-testid="CircleOutlinedIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`PrimaryIconButton tests should render the icon button with provided icon 1`] = `
<svg
  data-testid="test-icon"
/>
`;
